src/rmc.rs

```diff
@@ -0,0 +1,426 @@
+        signer_index: NodeIndex,
+        node_count: NodeCount,
+        keychain: &'a MK,
+        multisigned_hashes_tx: &UnboundedSender<Multisigned<'a, H, MK>>,
+    ) -> Option<Self> {
+        match maybe_state {
+            Some(HashState::Completed { .. }) => maybe_state,
+            Some(HashState::MultisignatureAggregation {
+                multisigned,
+                mut confirmed,
+                ..
+            }) => {
+                confirmed.set(signer_index);
+                if 3 * confirmed.true_indices().count() > 2 * node_count.0 {
+                    Some(HashState::Completed {
+                        _multisigned: multisigned,
+                    })
+                } else {
+                    Some(HashState::MultisignatureAggregation {
+                        multisigned,
+                        confirmed,
+                    })
+                }
+            }
+            _ => {
+                multisigned_hashes_tx
+                    .unbounded_send(multisigned_hash.clone())
+                    .expect("Send should succeed");
+                let mut confirmed = BoolNodeMap::with_capacity(node_count);
+                confirmed.set(signer_index);
+                confirmed.set(keychain.index());
+                Some(HashState::MultisignatureAggregation {
+                    multisigned: multisigned_hash,
+                    confirmed,
+                })
+            }
+        }
+    }
+
+    fn handle_unchecked_message(
+        maybe_state: Option<Self>,
+        message: IncomingMessage<H, MK>,
+        node_count: NodeCount,
+        keychain: &'a MK,
+        multisigned_hashes_tx: &UnboundedSender<Multisigned<'a, H, MK>>,
+    ) -> Option<Self> {
+        match message.message {
+            Message::SignedHash(unchecked) => {
+                let signed_indexed_hash = match unchecked.check(keychain) {
+                    Ok(signed_indexed_hash) => signed_indexed_hash,
+                    Err(_) => {
+                        debug!("Received a hash with a bad signature");
+                        return maybe_state;
+                    }
+                };
+                let indexed_hash = signed_indexed_hash.as_signable();
+                let index = indexed_hash.index();
+                if index != message.sender {
+                    debug!("Index of a message does not match the sender");
+                    // This is an unusual situation, but we can still process the hash because
+                    // it is signed by the node with the index from the message.
+                }
+                Self::handle_signed_hash(
+                    maybe_state,
+                    signed_indexed_hash,
+                    node_count,
+                    keychain,
+                    multisigned_hashes_tx,
+                )
+            }
+            Message::MultisignedHash(unchecked) => {
+                let signed_indexed_unchecked = match unchecked.check(keychain) {
```

Comment:
What would be an attack scenario with sending a valid multisignature that has invalid signature? The multisignature is implicitly signed by all partial signers so I believe that the additional signature is unnecessary and we could define `MultisignedHash(Indexed<UncheckedSigned<H, M>, S>)`.

Reply:
Replaced it with even simpler `MultisignedHash(UncheckedSigned<H, M>)`